### PR TITLE
feat: Compile workspace packages in parallel

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

This PR parallelizes compilation of packages in a workspace to reduce compilation times in large projects.

This currently only applies to the `nargo compile` command, we should roll this out across other commands in future.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

This is important as it allows us to remove a good fraction of the custom tooling in aztec-packages.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
This needs error reporting hoisted out.

**Comment by TomAFrench** [comment]
This is now complete. Just waiting on #2659 to be merged first.

File: tooling/nargo_cli/src/cli/compile_cmd.rs [hunk 1]
```diff
@@ -63,38 +64,53 @@ pub(crate) fn run(
...
+            .collect();
+    let contract_results: Vec<(
+        FileManager,
+        CompilationResult<Vec<(CompiledContract, DebugArtifact)>>,
+    )> = contract_packages
+        .par_iter()
+        .map(|package| {
+            compile_contracts(package, &args.compile_options, np_language, &is_opcode_supported)
+        })
+        .collect();
```
Comment (TomAFrench) [inline tooling/nargo_cli/src/cli/compile_cmd.rs:90]: Note that we still need to wait for all programs to compile before we compile contracts but this is a decent improvement on the current situation and we can iterate on this.

**Review by kevaundray — APPROVED** [review]
LGTM - I don't see anything else left to be done here?

**Comment by TomAFrench** [comment]
Nope, this is good to go in.

**Pull request merged**